docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+
+The key thing to remember about `any` is that it does not resolve errors, but only hides them. The errors still affect the code, while `any` makes it impossible to assess and counteract their influence.
+
+- `any` doesn't represent the widest type, or indeed any type at all. `any` is a compiler directive for _disabling_ static type checking for the value or type to which it's assigned.
+- `any` suppresses all error messages about its assignee. This makes code with `any` usage brittle against changes, since the compiler is unable to update its feedback even when the code has changed enough to alter or remove the error, or even add new type errors.
+- `any` subsumes all other types it comes into contact with. Any type that is in a union, intersection, is a property of, or has any other relationship with an `any` type or value becomes an `any` type itself. This represents an unmitigated loss of type information.
+
+```typescript
+// Type of 'payload_0': 'any'
+const handler:
+  | ((payload_0: ComposableControllerState, payload_1: Patch[]) => void)
+  | ((payload_0: any, payload_1: Patch[]) => void);
+
+function returnsAny(): any {
+  return { a: 1, b: true, c: 'c' };
+}
+// Types of a, b, c are all `any`
+const { a, b, c } = returnsAny();
+```
+
+- `any` infects all surrounding and downstream code with its directive to suppress errors. This is the most dangerous characteristic of `any`, as it causes the encroachment of unsafe code with no guarantees about type safety or runtime behavior.
+
+All of this makes `any` a prominent cause of dangerous **silent failures** (false negatives), where the code fails at runtime but the compiler does not provide any prior warning, which defeats the purpose of using a statically-typed language.
+
+##### When tempted to use `any`, try `unknown` and `never` instead
+
+###### `unknown`
+
+`any` usage is often motivated by a need to find a placeholder type that could be anything. `unknown` is the most likely type-safe substitute for `any` in these cases.
+
+- `unknown` is the universal supertype i.e. the widest possible type, equivalent to the universal set(U).
+- Every type is assignable to `unknown`, but `unknown` is only assignable to `unknown`.
+- When typing the _assignee_, `any` and `unknown` are completely interchangeable since every type is assignable to both.
+
+###### `never`
+
+`never` is worth trying as a starting point for narrowing down the type, as it is the universal subtype and assignable to all types.
+
+- `never` is the universal subtype i.e. the narrowest possible type, equivalent to the null set(∅).
+- `never` is assignable to every type, but the only type that is assignable to `never` is `never`.
+- When typing the _assigned_:
```

Comment:
What does "the assigned" mean here? I'm having trouble understanding this collection of points

Reply:
I wrote a short explanation about assignee and assigned with examples here: https://github.com/MetaMask/contributor-docs/issues/69#issuecomment-1982803504. 

I'd be happy to switch to a simpler pair of words, but there doesn't seem to be any official or popular terminology for this. I might add a glossary section to explain some of the opaque terminology I use within the article.

Reply:
Ah sorry, to clarify, I do understand what the terms mean. Just not how it's used in this context. Does "When typing the assigned" mean "When writing a type annotation"?

When this says "`unknown` is unable to replace `any`", does this mean that changing a type annotation from `any` to `unknown` can introduce errors?

I get that this situation is about explaining when `never` is an appropriate replacement for `any`, but I'm not grasping what the specific examples/scenarios are here.

Reply:
Yes `unknown` and `any` are only interchangeable if they're typing the assignee. 

If they're typing the assigned, it's likely that the `any` was used in the first place because the assigned type is too wide to fit into the assignee type "slot", resulting in an error. Using `unknown`, the widest type, as the substitute probably won't help. there. 

`never` will at least fit into the assignee, since it's assignable to every type.

Reply:
Hmm, I'm still not following how these points help with understanding when `never` can be used to replace `any`. If there is a broader point here, I am missing it.

Reply:
Well our `ActionConstraint['handler']` going from `(...args: any[]) => unknown` to `(...args: never[]) => unknown` is a good example, although function param contravariance is a bit of a special case.

Other than that, if the any-typed variable is going to be assigned to something else e.g. passed into another function as an argument, fixing it to `unknown` will not work (unless the assignee is also `unknown`), while `never` will always work. Here's an example:

```ts
// No error when `arg1` is `any` or `never`. 
// If `arg1` is `unknown`:
// Error: Argument of type 'unknown' is not assignable to parameter of type 'string'.(2345)
function f1(arg1: ???) {
  f2(arg1)
}
function f2(arg2: string) { ... }
```
https://tsplay.dev/mpodxN

Reply:
Hmm. That is a good example of why using `never` as a type annotation is problematic, but is there ever a reason to consider `never` here?

I've only seen `never` used personally in conditional types, or in generics. In conditional types its use is rather obvious, and could never be confused with `any`. So are generics the last case where we might expect to replace `any` with `never`? Or is there another case that I'm missing

Reply:
I think it would be clearer to rewrite this so that the distinction is between cases where `unknown` works as a substitute for `any` vs. cases where it doesn't. 

I agree that in the second case, it's not especially helpful to single out `never` as the first option to try just because it will almost always work.

In terms of `never` use cases, I have a draft for another entry: ["Prefer `as never` for empty types"](https://github.com/MetaMask/contributor-docs/issues/69#issuecomment-1910430960). But that isn't about replacing `any` so it doesn't apply here. 

It seems like the only `never` use case left to cover here is as a possible replacement for `any` in generic arguments.

Reply:
Applied here: https://github.com/MetaMask/contributor-docs/pull/74/commits/03bdb01d4210fa5fa0e24e92c672f3e2bdf6f8c0

Reply:
For what it's worth, the new changes are more approachable to me.